pkg/jobWorker/job.go

```diff
@@ -0,0 +1,413 @@
+		UUID:     uuid.New(),
+		config:   config,
+		output:   output,
+		exitCode: -1,
+	}
+	log.Printf("creted job:%s", job)
+	return job
+}
+
+// Start - starting the Job in a semi-isolated environment (creating new PID, mount and network and also creates a new control group for the process limiting CPU, IO, and memory)
+// The user running Start() should be the root user or have the necessary permissions to create namespaces and control groups
+//
+// ErrJobAlreadyStarted is returned, if the Job has already been started.
+// ErrInvalidCommand, ErrInvalidCPU, ErrInvalidIOBytesPerSecond, ErrInvalidMemBytes is returned, if provided configuration is invalid
+func (job *Job) Start() error {
+	job.mutex.Lock()
+	defer job.mutex.Unlock()
+
+	// validate configuration
+	log.Printf("validate job:%s", job)
+	if err := job.config.isValid(); err != nil {
+		log.Printf("validate job error:%v", err)
+		return err
+	}
+
+	// validate job hasn't been run
+	if job.isStarted {
+		return ErrJobAlreadyStarted
+	}
+
+	cmd := exec.Command(job.config.Command, job.config.Arguments...)
+	// combine the stdout and stderr so that the stdout and stderr are combined in the order they are written
+	cmd.Stderr = job.output
+	cmd.Stdout = job.output
+
+	cmd.SysProcAttr = &syscall.SysProcAttr{
+		// CLONE_NEWPID:  creates a new PID namespace preventing the process from seeing/killing host processes
+		// CLONE_NEWNET:  creates a new network namespace preventing the process from accessing the internet or local network
+		// CLONE_NEWNS:   creates a new mount namespace preventing the process from impacting host mounts
+		// CLONE_NEWUTS:  creates a new UTS namespaces provide isolation between two system identifiers: the hostname and the NIS domain name
+		// CLONE_NEWUSER: creates new namespaces to isolate security-related identifiers and attributes, in particular, user IDs and group IDs
+		Cloneflags: syscall.CLONE_NEWNS |
+			//	syscall.CLONE_NEWIPC | |
+			syscall.CLONE_NEWNET |
+			syscall.CLONE_NEWUTS |
+			syscall.CLONE_NEWPID |
+			syscall.CLONE_NEWUSER,
+		UidMappings: []syscall.SysProcIDMap{
+			{
+				ContainerID: 0,
+				HostID:      os.Getuid(),
+				Size:        1,
+			},
+		},
+		GidMappings: []syscall.SysProcIDMap{
+			{
+				ContainerID: 0,
+				HostID:      os.Getgid(),
+				Size:        1,
+			},
+		},
+		// force the child processes to start in theirs own process groups
+		Setsid: true,
+		Pgid:   0,
+		//	// Also, enables mounting a new proc filesystem so that command such as `ps -ef` only see the processes in the PID namespace
+		//	Unshareflags: syscall.CLONE_NEWNS,
+		//	// instruct cmd.Run to use the control group file descriptor, so that Job Command does not
+		//	// have to manually add the new PID to the control group
+		//	UseCgroupFD: true,
+	}
+
+	cleanCGroup := make(chan bool)
+	waitingCleanCGroupToCompleted := func() {
+		_ = <-cleanCGroup // send notification channel cleaned up
+	}
+
+	go func() {
+		time.Sleep(100 * time.Millisecond)
```

Comment:
This is dirty approach where the time Sleep is used to artificially delay operation to force  particular condition for UT.

Reply:
Sorry, it was done for testing purpose to prove this: https://github.com/P-A-R-U-S/Go-Job-Worker-Service/pull/2#discussion_r1747710054 and I've removed `sleep`.